### PR TITLE
Exception Handling and Examples for Expected Failures

### PR DESCRIPTION
**Pull request by dvonthenen**
Addresses issue: https://github.com/deepgram/deepgram-dotnet-sdk/issues/252

This was the stress test to throw as many exceptions and to test edge cases as best as I could. In these scenarios, I found a few issues.

Changes include:
- In `AbstractRestClient` just better logging of the exceptions based on log levels
- In `LiveClient`, this is where I found a couple of issues went it came to cancelling. These changes include:
  - `Connect()` and `Stop()` Each get their own CancelToken to be used for only that operation. That is the token that is passed into the function. That token is optional and a default will be used to timeout those operations in 5 seconds.
  - The internal CancelToken is only used for cancelling out of threads for thread management. I initially try to see if I could compact these two into a single token, but that proves to be "janky" and unpredictable. KISS.
  - Fixes exception logging similar to that of `AbstractRestClient`.
- Implemented the ability to disable Logging
- Cleaned up some examples for Demo purposes. 

## Summary by CodeRabbit

- **New Features**
	- Added new projects for handling different scenarios such as trigger cancellation, bad host detection, and WebSocket management.
	- Introduced a default timeout for connect/disconnect operations to improve reliability.
	- Implemented a new logging level to allow disabling logs, enhancing user control over application verbosity.
- **Improvements**
	- Enhanced logging statements across the application for better clarity and debugging.
	- Improved exception handling messages to include more detailed information.
	- Updated WebSocket closure logic for more robust connection management.
- **Documentation**
	- Added examples and tests showcasing the usage of the SDK for microphone streaming and handling specific failure scenarios.
- **Bug Fixes**
	- Corrected project reference paths to resolve file path issues in test projects.

### CODE REVIEW

File: Deepgram/Logger/Log.cs [hunk 1]
```diff
@@ -21,17 +22,25 @@ static Log()
...
     public static Serilog.ILogger Initialize(LogLevel level = LogLevel.Information, string? filename = "log.txt")
     {
+        if (level == LogLevel.Disable)
+        {
+            instance = new LoggerConfiguration()
+                .MinimumLevel.Is(((LogEventLevel)1 + (int)LogEventLevel.Fatal))
+                .WriteTo.Console(outputTemplate: "{Timestamp:yyyy-MM-dd HH:mm:ss.fff} [{Level}] {Message}{NewLine}{Exception}")
+                .CreateLogger();
+            return instance;
+        }
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Logger/Log.cs:32]: That is totally wrong. The suggestion above would log fatal errors. We want to disable it all. Ignoring suggestion.

File: Deepgram/Abstractions/AbstractRestClient.cs [hunk 2]
```diff
@@ -83,21 +83,29 @@ internal AbstractRestClient(string? apiKey = null, DeepgramHttpClientOptions? op
...
+            Log.Verbose("AbstractRestClient.GetAsync<T>", "LEAVE");
+            throw;
+        }
         catch (Exception ex)
         {
-            Log.Error("GetAsync<T>", $"Exception: {ex}");
-            Log.Verbose("GetAsync<T>", "LEAVE");
+            Log.Error("GetAsync<T>", $"{ex.GetType()} thrown {ex.Message}");
+            Log.Verbose("GetAsync<T>", $"Excepton: {ex}");
+            Log.Verbose("AbstractRestClient.GetAsync<T>", "LEAVE");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Abstractions/AbstractRestClient.cs:100]: That's what the second `Verbose` line is for. We only want the error message in the default logging level. Ignoring suggestion.

File: Deepgram/Clients/Live/v1/Client.cs [hunk 3]
```diff
@@ -663,110 +671,110 @@
...
             Log.Verbose("Stop", $"Stop cancelled. Info: {ex}");
+            Log.Verbose("LiveClient.Stop", "LEAVE");
         }
         catch (Exception ex)
         {
-            Log.Error("Stop", $"Excepton: {ex}");
+            Log.Error("Stop", $"{ex.GetType()} thrown {ex.Message}");
+            Log.Verbose("Stop", $"Excepton: {ex}");
+            Log.Verbose("LiveClient.Stop", "LEAVE");
+            throw;
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:776]: nothing is swallowed. All exceptions are re-thrown.

File: Deepgram/Clients/Live/v1/Client.cs [hunk 4]
```diff
@@ -663,110 +671,110 @@
...
+            Log.Error("ProcessDataReceived", $"{ex.GetType()} thrown {ex.Message}");
+            Log.Verbose("ProcessDataReceived", $"Excepton: {ex}");
+            Log.Verbose("LiveClient.ProcessDataReceived", "LEAVE");
         }
         catch (Exception ex)
         {
-            Log.Error("ProcessDataReceived", $"Excepton: {ex}");
+            Log.Error("ProcessDataReceived", $"{ex.GetType()} thrown {ex.Message}");
+            Log.Verbose("ProcessDataReceived", $"Excepton: {ex}");
+            Log.Verbose("LiveClient.ProcessDataReceived", "LEAVE");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:686]: That's what the second `Verbose` line is for. We only want the error message in the default logging level. Ignoring suggestion.

File: Deepgram/Clients/Live/v1/Client.cs [hunk 5]
```diff
@@ -145,10 +144,20 @@ public Client(string? apiKey = null, DeepgramWsClientOptions? options = null)
...
+            Log.Verbose("Connect", $"Connect cancelled. Info: {ex}");
+            Log.Verbose("LiveClient.Connect", "LEAVE");
         }
         catch (Exception ex)
         {
-            Log.Error("Connect", $"Excepton: {ex}");
+            Log.Error("Connect", $"{ex.GetType()} thrown {ex.Message}");
+            Log.Verbose("Connect", $"Excepton: {ex}");
+            Log.Verbose("LiveClient.Connect", "LEAVE");
+            throw;
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:160]: That's what the second `Verbose` line is for. We only want the error message in the default logging level. Ignoring suggestion.

File: Deepgram/Abstractions/AbstractRestClient.cs [hunk 6]
```diff
@@ -132,14 +140,22 @@
...
+            Log.Verbose("AbstractRestClient.GetAsync<S, T>", "LEAVE");
+            throw;
+        }
         catch (Exception ex)
         {
-            Log.Error("GetAsync<S, T>", $"Exception: {ex}");
-            Log.Verbose("GetAsync<S, T>", "LEAVE");
+            Log.Error("GetAsync<S, T>", $"{ex.GetType()} thrown {ex.Message}");
+            Log.Verbose("GetAsync<S, T>", $"Excepton: {ex}");
+            Log.Verbose("AbstractRestClient.GetAsync<S, T>", "LEAVE");
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Abstractions/AbstractRestClient.cs:158]: I think it's pretty straight forward enough. Ignoring suggestion.

File: Deepgram/Clients/Live/v1/Client.cs [hunk 7]
```diff
@@ -663,110 +671,110 @@
...
             return;
         }
 
-        // send the close message and flush transcription message
-        try {
+        if (cancelToken == null)
+        {
+            Log.Information("Stop", "Using default disconnect cancellation token");
+            cancelToken = new CancellationTokenSource(Constants.DefaultConnectTimeout);
+        }
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:710]: Will update!

File: Deepgram/Clients/Live/v1/Client.cs [hunk 8]
```diff
@@ -663,110 +671,110 @@
...
+            // if websocket is open, send a close message
             if (_clientWebSocket!.State == WebSocketState.Open)
             {
                 Log.Debug("Stop", "Sending Close message...");
-
                 // send a close to Deepgram
                 lock (_mutexSend)
                 {
-                    _clientWebSocket.SendAsync(new ArraySegment<byte>([]), WebSocketMessageType.Binary, true, cancelToken)
+                    _clientWebSocket.SendAsync(new ArraySegment<byte>([]), WebSocketMessageType.Binary, true, cancelToken.Token)
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:728]: Will update!

File: Deepgram/Clients/Live/v1/Client.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:710]: Actually, this isn't needed. We need to create the cancel token for `Stop()` because this handles any calls to functions within `Stop()` that might actually need to be called should the socket be open. Ignoring suggestion.

**Review by jpvajda — DISMISSED** [review]
LGTM

**Review by jpvajda — APPROVED** [review]
LGTM
